### PR TITLE
Stat class file closing

### PR DESCRIPTION
**Pull request by GarethCabournDavies**
This should solve issue #4116; basically doing as suggested, only saving the filenames and wrapping each call to open the file in a `with` statement (except one case which explicitly `close()`s the file)

I think this is good for the idq implementation as well, so hopefully @dethodav / @maxtrevor can comment on that

Testing running now, but the examples don't have iDQ in them so it won't test all parts of the changes

### CODE REVIEW

File: pycbc/events/stat.py [hunk 1]
```diff
@@ -1240,12 +1247,14 @@ def reassign_rate(self, ifo):
...
+        with h5py.File(self.files[f'{ifo}-fit_coeffs'], 'r') as coeff_file:
+            analysis_time = float(coeff_file.attrs['analysis_time'])
+            fbt = 'fit_by_template' in coeff_file
+
         self.fits_by_tid[ifo]['smoothed_rate_above_thresh'] /= analysis_time
         self.fits_by_tid[ifo]['smoothed_rate_in_template'] /= analysis_time
         # The by-template fits may have been stored in the smoothed fits file
-        if 'fit_by_template' in coeff_file:
+        if fbt:
             self.fits_by_tid[ifo]['fit_by_rate_above_thresh'] /= analysis_time
```
Comment (tdent) [inline pycbc/events/stat.py:1258]: For my edification, where are these by-template fit values used?

Comment (GarethCabournDavies) [reply pycbc/events/stat.py:1258]: They are used in the cuts; cutting on fit_by values fixed a significant problem in the singles search with isolated templates having lots of triggers, and then that being smoothed over nearby templates and rates are then underestimated

Comment (GarethCabournDavies) [reply pycbc/events/stat.py:1258]: So https://github.com/gwastro/pycbc/blob/master/pycbc/events/cuts.py#L44 and https://github.com/gwastro/pycbc/blob/master/pycbc/events/cuts.py#L206

**Review by spxiwh — APPROVED** [review]
This all looks like no behaviour is changed by any of this, other than closing files as intended. Thanks @GarethCabournDavies !

**Comment by spxiwh** [comment]
Given that this affects the search workflow though, the search test *must* complete before merging.

**Comment by GarethCabournDavies** [comment]
Having got back to this, I think the search now works properly, but the test is failing for another reason

**Pull request merged**